ic-factory/src/top_up.rs

```diff
@@ -0,0 +1,148 @@
+use candid::Principal;
+use cycles_minting_canister::{
+    IcpXdrConversionRateCertifiedResponse, NotifyError, NotifyTopUp, DEFAULT_CYCLES_PER_XDR,
+    MEMO_TOP_UP_CANISTER,
+};
+use ic_base_types::{CanisterId, PrincipalId};
+use ic_canister::virtual_canister_call;
+use ledger_canister::{
+    AccountIdentifier, BlockHeight, SendArgs, Subaccount, Tokens, DEFAULT_TRANSFER_FEE,
+    TOKEN_SUBDIVIDABLE_BY,
+};
+use num_traits::ToPrimitive;
+
+use crate::error::FactoryError;
+
+const CYCLE_MINTING_CANISTER: &str = "rkp4c-7iaaa-aaaaa-aaaca-cai";
+
+/// This function calculates the amount required for minting cycles for a canister.
+pub async fn cycles_to_icp(cycles: u64) -> Result<u64, FactoryError> {
+    let rate = get_conversion_rate().await?.data;
+
+    let icp_per_xdr = 10_000.0 / rate.xdr_permyriad_per_icp as f64;
```

Comment:
We don't have to use f64 here (in this whole function, not only this line). If you first do all multiplications and then all divisions, unsigned integers would work just fine. Also, check for division by 0 is necessary here I believe.